File: src/rename/proximity.test.ts
import assert from "node:assert";
import { createIsEligible } from "./rename-eligibility.js";
import { describe, it } from "node:test";
import { getProximateUsedNames } from "./proximity.js";

describe("getProximateUsedNames", () => {
  function makeBinding(line: number, refLines: number[] = []) {
    return {
      identifier: { loc: { start: { line } } },
      referencePaths: refLines.map((l) => ({
        node: { loc: { start: { line: l } } }
      }))
    };
  }

  it("always includes well-known names", () => {
    const allNames = new Set(["exports", "require", "console", "a", "b"]);
    const scopeBindings: Record<string, ReturnType<typeof makeBinding>> = {
      exports: makeBinding(1),
      require: makeBinding(2),
      console: makeBinding(3),
      a: makeBinding(1000), // far away
      b: makeBinding(1001) // far away
    };

    const result = getProximateUsedNames(
      allNames,
      [50],
      scopeBindings,
      200,
      createIsEligible()
    );

    assert.ok(result.has("exports"), "should include well-known 'exports'");
    assert.ok(result.has("require"), "should include well-known 'require'");
    assert.ok(result.has("console"), "should include well-known 'console'");
  });

  it("excludes eligible names", () => {
    // With the default isEligible, single-char names and descriptive names
    // are all eligible (everything is a rename candidate). Use an override
    // that treats only single-char names as eligible.
    const isEligible = (name: string) => name.length === 1;
    const allNames = new Set(["a", "b", "c", "myVar"]);
    const scopeBindings: Record<string, ReturnType<typeof makeBinding>> = {
      a: makeBinding(50),
      b: makeBinding(50),
      c: makeBinding(50),
      myVar: makeBinding(50)
    };

    const result = getProximateUsedNames(
      allNames,
      [50],
      scopeBindings,
      200,
      isEligible
    );

    assert.ok(!result.has("a"), "should exclude eligible 'a'");
    assert.ok(!result.has("b"), "should exclude eligible 'b'");
    assert.ok(!result.has("c"), "should exclude eligible 'c'");
    assert.ok(result.has("myVar"), "should include non-eligible 'myVar'");
  });

  it("includes names within +-100 lines, excludes those outside", () => {
    // Use an override that treats only single-char names as eligible,
    // so nearVar/farVar are preserved and subject to windowing
    const isEligible = (name: string) => name.length === 1;
    const allNames = new Set(["nearVar", "farVar"]);
    const scopeBindings: Record<string, ReturnType<typeof makeBinding>> = {
      nearVar: makeBinding(55), // within +-100 of line 50
      farVar: makeBinding(500) // far away from line 50
    };

    const result = getProximateUsedNames(
      allNames,
      [50],
      scopeBindings,
      200,
      isEligible
    );

    assert.ok(result.has("nearVar"), "should include name within proximity");
    assert.ok(!result.has("farVar"), "should exclude name outside proximity");
  });

  it("includes name if any reference is within proximity", () => {
    const isEligible = (name: string) => name.length === 1;
    const allNames = new Set(["refVar"]);
    const scopeBindings: Record<string, ReturnType<typeof makeBinding>> = {
      refVar: makeBinding(500, [45]) // declaration far, but reference near line 50
    };

    const result = getProximateUsedNames(
      allNames,
      [50],
      scopeBindings,
      200,
      isEligible
    );

    assert.ok(
      result.has("refVar"),
      "should include name whose reference is within proximity"
    );
  });

  it("returns all preserved names when below threshold", () => {
    const isEligible = (name: string) => name.length === 1;
    const allNames = new Set(["nearVar", "farVar", "a"]);
    const scopeBindings: Record<string, ReturnType<typeof makeBinding>> = {
      nearVar: makeBinding(50),
      farVar: makeBinding(500),
      a: makeBinding(50)
    };

    // totalBindings < 100 -> no windowing
    const result = getProximateUsedNames(
      allNames,
      [50],
      scopeBindings,
      50,
      isEligible
    );

    assert.ok(result.has("nearVar"), "should include nearVar");
    assert.ok(
      result.has("farVar"),
      "should include farVar (no windowing below threshold)"
    );
    assert.ok(!result.has("a"), "should still exclude eligible names");
  });
});

/**
 * 16-findings-queue #22 (finding #12's class, found by the Rust port,
 * WP3.3): `scopeBindings[name]` fell through to Object.prototype, so a used
 * name like `toString` (the LLM assigns such names) that is NOT bound in
 * this scope read the built-in function — truthy, no loc, no references —
 * and was EXCLUDED from the proximate used names, where a genuinely absent
 * name is included ("include if binding not found, to be safe").
 */
describe("getProximateUsedNames reads only own scope bindings", () => {
  it("includes an unbound toString like any other unbound name", () => {
    const scopeBindings: Record<
      string,
      { identifier: { loc: { start: { line: number } } } }
    > = {};
    for (let i = 0; i < 200; i++) {
      scopeBindings[`far${i}`] = {
        identifier: { loc: { start: { line: 9000 + i } } }
      };
    }
    const result = getProximateUsedNames(
      new Set(["toString", "someUnboundName"]),
      [50],
      scopeBindings,
      200,
      (name: string) => name.length === 1
    );
    assert.ok(result.has("someUnboundName"), "an unbound name is included");
    assert.ok(
      result.has("toString"),
      "an unbound toString must be included too"
    );
  });
});

File: crates/humanify-core/src/rename/votes/proximity/proximity_test.rs
//! TS original: `src/rename/proximity.test.ts` (its five cases), plus the
//! inherited-record case the TS carries by accident.

use std::collections::HashMap;

use super::{ProximityBinding, get_proximate_used_names};

fn binding(line: u32, refs: &[u32]) -> ProximityBinding {
    ProximityBinding {
        decl_line: Some(line),
        ref_lines: refs.to_vec(),
    }
}

fn names(list: &[&str]) -> Vec<String> {
    list.iter().map(|s| s.to_string()).collect()
}

fn run(
    all: &[&str],
    lines: &[u32],
    bindings: &HashMap<&str, ProximityBinding>,
    total: usize,
    eligible: impl Fn(&str) -> bool,
) -> Vec<String> {
    get_proximate_used_names(
        &names(all),
        lines,
        |n| bindings.get(n).cloned(),
        total,
        eligible,
    )
}

#[test]
fn always_includes_well_known_names() {
    let b = HashMap::from([
        ("exports", binding(1, &[])),
        ("require", binding(2, &[])),
        ("console", binding(3, &[])),
        ("a", binding(1000, &[])),
        ("b", binding(1001, &[])),
    ]);
    let out = run(
        &["exports", "require", "console", "a", "b"],
        &[50],
        &b,
        200,
        |_| true,
    );
    for n in ["exports", "require", "console"] {
        assert!(out.iter().any(|x| x == n), "{n}");
    }
}

#[test]
fn excludes_eligible_names() {
    let b = HashMap::from([
        ("a", binding(50, &[])),
        ("b", binding(50, &[])),
        ("c", binding(50, &[])),
        ("myVar", binding(50, &[])),
    ]);
    let out = run(&["a", "b", "c", "myVar"], &[50], &b, 200, |n| n.len() == 1);
    assert_eq!(out, names(&["myVar"]));
}

#[test]
fn includes_names_within_the_radius_only() {
    let b = HashMap::from([("nearVar", binding(55, &[])), ("farVar", binding(500, &[]))]);
    let out = run(&["nearVar", "farVar"], &[50], &b, 200, |n| n.len() == 1);
    assert_eq!(out, names(&["nearVar"]));
}

#[test]
fn includes_a_name_whose_reference_is_near() {
    let b = HashMap::from([("refVar", binding(500, &[45]))]);
    let out = run(&["refVar"], &[50], &b, 200, |n| n.len() == 1);
    assert_eq!(out, names(&["refVar"]));
}

#[test]
fn returns_every_preserved_name_below_the_threshold() {
    let b = HashMap::from([
        ("nearVar", binding(50, &[])),
        ("farVar", binding(500, &[])),
        ("a", binding(50, &[])),
    ]);
    let out = run(&["nearVar", "farVar", "a"], &[50], &b, 50, |n| n.len() == 1);
    assert_eq!(out, names(&["nearVar", "farVar"]));
}

#[test]
fn an_absent_binding_is_included_but_an_inherited_record_name_is_not() {
    // `scopeBindings[name]` on a plain object: a truly absent name reads
    // undefined (included "to be safe"); `toString` reads the inherited
    // Object.prototype function — truthy, no loc — and is excluded.
    let b: HashMap<&str, ProximityBinding> = HashMap::new();
    let out = run(&["missingVar", "toString"], &[50], &b, 200, |_| false);
    assert_eq!(out, names(&["missingVar"]));
}
